pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,421 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/api/admissionregistration/v1alpha1"
+	"k8s.io/api/core/v1"
+	clientretry "k8s.io/kubernetes/pkg/client/retry"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	"k8s.io/kubernetes/pkg/controller"
+	kubeletapis "k8s.io/kubernetes/pkg/kubelet/apis"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+var updatePVLabelsBackoff = wait.Backoff{
+	Steps:    20,
+	Duration: 50 * time.Millisecond,
+	Jitter:   1.0,
+}
+
+const initializerName = "pvlabel.kubernetes.io"
+const initialzerConfigName = "pvlabel-initconfig"
+
+// PersistentVolumeLabelController handles adding labels to persistent volumes when they are created
+type PersistentVolumeLabelController struct {
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
+	cloud            cloudprovider.Interface
+	mutex            sync.Mutex
+	kubeClient       kubernetes.Interface
+	pvlController    cache.Controller
+
+	syncHandler func(vol *v1.PersistentVolume) error
+
+	// queue is where incoming work is placed to de-dup and to allow "easy" rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+}
+
+// NewPersistentVolumeLabelController creates a PersistentVolumeLabelController object
+func NewPersistentVolumeLabelController(
+	kubeClient kubernetes.Interface,
+	cloud cloudprovider.Interface) *PersistentVolumeLabelController {
+	pvlc := &PersistentVolumeLabelController{
+		cloud:      cloud,
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pvLabels"),
+	}
+	pvlc.syncHandler = pvlc.AddLabels
+	_, pvlc.pvlController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.IncludeUninitialized = true
+				return kubeClient.CoreV1().PersistentVolumes().Watch(options)
+			},
+		},
+		&v1.PersistentVolume{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cast := obj.(*v1.PersistentVolume)
+				pvlc.queue.Add(cast)
+			},
+		},
+	)
+	//	pvlc.pvSynced = pvlc.pvlController.HasSynced
+
+	return pvlc
+}
+
+// Run starts a controller that adds labels to persistent volumes
+func (pvlc *PersistentVolumeLabelController) Run(threadiness int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer pvlc.queue.ShutDown()
+
+	glog.Infof("Starting PersistentVolumeLabelController")
+	defer glog.Infof("Shutting down PersistentVolumeLabelController")
+
+	go pvlc.pvlController.Run(stopCh)
+
+	if !controller.WaitForCacheSync("persistent volume label", stopCh, pvlc.pvlController.HasSynced) {
+		return
+	}
+
+	// add an initializer
+	pvlc.addInitializationConfig()
```

Comment:
I also think self-registration seems a little strange. An initializer controller and the registration mechanics under a certain name/scope seem to be separate concerns. I expected the controller to be created, started, health-assessed, and *then* registered in `controllermanager.go`. Separation allows for more control over registration itself (e.g. making it optional or otherwise configurable in a context where such configuration is available).